library/std/src/io/buffered/bufwriter.rs

```diff
@@ -291,55 +291,94 @@ impl<W: Write> BufWriter<W> {
 #[stable(feature = "rust1", since = "1.0.0")]
 impl<W: Write> Write for BufWriter<W> {
     fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        // We assume that callers of `write` prefer to avoid split writes where
+        // possible, so we pre-flush the buffer rather than doing a partial
+        // write to fill it.
         if self.buf.len() + buf.len() > self.buf.capacity() {
             self.flush_buf()?;
         }
-        // FIXME: Why no len > capacity? Why not buffer len == capacity? #72919
         if buf.len() >= self.buf.capacity() {
-            self.panicked = true;
-            let r = self.get_mut().write(buf);
-            self.panicked = false;
-            r
+            self.get_mut().write(buf)
         } else {
             self.buf.extend_from_slice(buf);
             Ok(buf.len())
         }
     }
 
     fn write_all(&mut self, buf: &[u8]) -> io::Result<()> {
-        // Normally, `write_all` just calls `write` in a loop. We can do better
-        // by calling `self.get_mut().write_all()` directly, which avoids
-        // round trips through the buffer in the event of a series of partial
-        // writes in some circumstances.
-        if self.buf.len() + buf.len() > self.buf.capacity() {
+        // Unlike with `write`, we assume that a caller of `write_all` is
+        // interested in minimizing system calls even if the buffer is split.
+        // This method tries to fill up the buffer as much as possible before
+        // flushing, whereas `write` prefers not split incoming bufs.
+
+        // Bypass the buffer if the the incoming write is larger than the whole
+        // buffer.
+        if buf.len() >= self.capacity() {
             self.flush_buf()?;
+            return self.get_mut().write_all(buf);
```

Comment:
couldn't this use a vectored write (if available) to write out both buffers together to further reduce syscalls?

Reply:
I looked into this a while ago (in fact, it was your suggestion back then! https://github.com/rust-lang/rust/pull/72808#discussion_r434101103).

This is possible (and, in fact, it actually makes sense to do it unconditionally because it's "no worse" than manually doing a pair of writes). It just turns out to be nontrivial, because you need to make sure you do the correct thing at the boundary of buffered and new data; you need to make sure to retry writing buffered data but (maybe) not retry writing new data. However, this PR is a great place to give it a try, so I'll see about updating `write` & `write_all` to make use of this.

Reply:
In what ended up being a very long and surprisingly enjoyable refactor, `BufWriter` now takes advantage of vectored writes when it makes sense to do so! The key new method is `flush_buf_vectored`, which is the same as `flush_buf` but will also attempt to send *new* data via a vectored write during the flush. Critically, it only continues writing while there is still unflushed data in the buffer; once the buffer is empty (even if none of the *new* data has been sent) it will return immediately. I did my best to comment thoroughly this method and its usages throughout `BufWriter`.

Reply:
I also introduced *some* optimizations to `LineWriterShim` along the same lines (taking advantage of `flush_buf_vectored`. There may be more opportunity there, but I'm considering a larger refactor of line writing to be out of scope for this PR.